File: services/intelligence/app/models/security.py
from sqlalchemy import JSON, Date, DateTime, String
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class Security(Base):
    __tablename__ = "securities"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    canonical_symbol: Mapped[str] = mapped_column(String(32), unique=True, nullable=False)
    symbol: Mapped[str] = mapped_column(String(16), nullable=False)
    exchange: Mapped[str] = mapped_column(String(16), nullable=False)
    market: Mapped[str] = mapped_column(String(16), default="CN", nullable=False)
    security_type: Mapped[str] = mapped_column(String(32), default="stock", nullable=False)
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    short_name: Mapped[str | None] = mapped_column(String(64), nullable=True)
    pinyin: Mapped[str | None] = mapped_column(String(128), nullable=True)
    list_date: Mapped[Date | None] = mapped_column(Date, nullable=True)
    delist_date: Mapped[Date | None] = mapped_column(Date, nullable=True)
    status: Mapped[str] = mapped_column(String(32), default="active", nullable=False)
    currency: Mapped[str] = mapped_column(String(8), default="CNY", nullable=False)
    metadata_json: Mapped[dict | None] = mapped_column("metadata", JSON, nullable=True)
    created_at: Mapped[DateTime | None] = mapped_column(DateTime, nullable=True)
    updated_at: Mapped[DateTime | None] = mapped_column(DateTime, nullable=True)

File: apps/api/app/Models/Security.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Security extends Model
{
    use HasFactory;

    protected $fillable = [
        'canonical_symbol',
        'symbol',
        'exchange',
        'market',
        'security_type',
        'name',
        'short_name',
        'pinyin',
        'list_date',
        'delist_date',
        'status',
        'currency',
        'metadata',
    ];

    protected $casts = [
        'metadata' => 'array',
        'list_date' => 'date',
        'delist_date' => 'date',
    ];

    public function articles(): BelongsToMany
    {
        return $this->belongsToMany(NewsArticle::class, 'news_article_securities', 'security_id', 'news_article_id')
            ->withPivot(['mention_type', 'matched_text', 'confidence', 'is_primary'])
            ->withTimestamps();
    }

    public function events(): HasMany
    {
        return $this->hasMany(MarketEvent::class, 'primary_security_id');
    }

    public function signals(): HasMany
    {
        return $this->hasMany(TradingSignal::class, 'primary_security_id');
    }
}
